[Tay Yan Han]/iP

tyanhan opened the pull request:
### Saitama
>"OK..." - Saitama

Saitama frees your mind of having to remember things you need to do. It's,
- text-based
- easy to learn
- ~~FAST~~ SUPER FAST to use

All you need to do is,

1. download it from [here](https://github.com/tyanhan/ip/releases)
2. double-click it.
3. add your tasks.
4. let Saitama manage your tasks for you! 👊 

And it is **FREE**!

Features:

- [x] Manage tasks
- [x] Manage deadlines
- [ ] Reminders (coming soon)

If you Java programmer, you can use it to practice Java too. Here's the `main` method:
```java
public static void main(String[] args) {
        new Saitama(filePath).run();
}
```

Commit:
Add Gradle support

tyanhan committed:
Add Greet and Echo

tyanhan committed:
Add list and add functionality

tyanhan committed:
Add Mark as Done

tyanhan committed:
Add Deadline, Event and Task functionality

tyanhan committed:
Change case sensitivity

tyanhan committed:
Add EmptyDescriptionException

tyanhan committed:
Add InvalidCommandException

tyanhan committed:
Add InvalidTaskNumberException

tyanhan committed:
Add InvalidTaskNumberException

tyanhan committed:
Convert if-else to switch statement

tyanhan committed:
Change dialog

tyanhan committed:
Change dialog

tyanhan committed:
Test new branch

tyanhan committed:
Change dialog

tyanhan committed:
Merge branch 'master' into feature1

tyanhan committed:
Remove TEST

tyanhan committed:
Change dialog

tyanhan committed:
Merge branch 'feature1'

tyanhan committed:
Add country

tyanhan committed:
Test branch

tyanhan committed:
Revert "Add country"

This reverts commit e6415af7e118d4ecf4564d81f413b1d538d129b7.

tyanhan committed:
Add dialog

tyanhan committed:
Merge branch 'add-countries'

tyanhan committed:
Add enum

tyanhan committed:
Fix add event bug

tyanhan committed:
Test feature 1

tyanhan committed:
Merge branch 'master' into feature1

# Conflicts:
#	src/main/java/Duke.java

tyanhan committed:
Remove feature 1

tyanhan committed:
Add save and load functionality

tyanhan committed:
Add Deadline read dates functionality

tyanhan committed:
Merge branch 'branch-Level-7'

tyanhan committed:
Merge branch 'branch-Level-8'

tyanhan committed:
Add Ui, Storage, Parser

tyanhan committed:
Package files into corresponding packages

tyanhan committed:
Add JavaDoc

tyanhan committed:
Update code style

tyanhan committed:
Add JavaDoc for Ui class

tyanhan committed:
Add Find to search task

tyanhan committed:
Merge branch 'branch-A-JavaDoc'

tyanhan committed:
Merge branch 'branch-A-CodingStandard'

tyanhan committed:
Merge branch 'branch-Level-9'

tyanhan committed:
Add JUnit tests and moved dialog from tasklist to Command

tyanhan committed:
Add JavaDoc for tasks

tyanhan committed:
Add fallthrough comment to switch

tyanhan committed:
Merge branch 'add-gradle-support' of https://github.com/tyanhan/ip into add-gradle-support

tyanhan committed:
Merge branch 'add-gradle-support'

tyanhan committed:
Add gradle support and correct styling errors

tyanhan committed:
Fix code style violations and add FileCorruptException

tyanhan committed:
Add GUI for Saitama

tyanhan committed:
Add JavaDocs

tyanhan committed:
Merge branch 'branch-Level-10'

tyanhan committed:
Add welcome dialog and fix dialog box bug

tyanhan committed:
Add wrapping to dialog box text

KwanHW reviewed (commented):
Hi Yan Han!

Well done on your codestyle and naming of variables/functions. The naming was consistent and easy to follow.

Some nits you can do:
- Your compiled java classes are not ignored in the repository. You can do so by the steps below
1. Add this line `*.class` into your `.gitignore` file
2. [Untrack](https://www.codeblocq.com/2016/01/Untrack-files-already-added-to-git-repository-based-on-gitignore/) the existing `.class` files using the attached URL

SethCKL commented on src/main/java/saitama/parser/Parser.java:
```diff
@@ -0,0 +1,192 @@
+        splitCommand[0] = splitCommand[0].toUpperCase();
+        String command = splitCommand[0];
+
+        switch (command) {
+        case "BYE":
+            return new ExitCommand();
+        case "LIST":
+            return new ListCommand();
+        case "MARK":
+            return prepareMark(splitCommand);
+        case "UNMARK":
+            return prepareUnmark(splitCommand);
+        case "DELETE":
+            return prepareDelete(splitCommand);
+        case "FIND":
+            return prepareFind(splitCommand);
+        case "TODO":
+            //Fallthrough
+        case "DEADLINE":
+            //Fallthrough
+        case "EVENT":
+            //Fallthrough
+            return prepareAdd(splitCommand);
+        default:
+            throw new InvalidCommandException();
```
Really like the choice to abstract every single command into it's own function.

Makes the code very readable and easy to maintain going forward should one choose to change the functionality or add new commands.

LGTM! code is clean and follows the coding standards

chriswengqi replied:
Agree with Seth! Really liked the idea of abstracting every single command into its own function in its respective class. As a reader I feel comfortable in reading your code.

chriswengqi commented on src/main/java/saitama/tasks/TaskList.java:
```diff
@@ -0,0 +1,114 @@
+     * @return The string format of a task list.
+     */
+    @Override
+    public String toString() {
+        int counter = 1;
+        String output = "";
+        for (Task task : taskList) {
+            output += counter + "." + task + "\n";
+            counter += 1;
+        }
+        return output.substring(0, output.length() - 1);
+    }
```
LGTM! Would suggest that you use `StringBuilder` instead of string concatenation `+=` for better efficiency in large test cases.

chriswengqi reviewed (approved):
LGTM! I learned a lot from reading your code :)

tyanhan committed:
Change profile picture to circle

tyanhan committed:
Add assert statements

tyanhan committed:
Merge pull request #1 from tyanhan/A-Assertions

Add assert statements

tyanhan committed:
Improve code quality

tyanhan committed:
Add Continuous Integration

tyanhan committed:
Merge pull request #2 from tyanhan/branch-A-CodeQuality

Improve code quality

tyanhan committed:
Merge pull request #3 from tyanhan/branch-A-CI

Add Continuous Integration

tyanhan committed:
Add time support for deadline tasks

The deadline task does not track time, only date.

Tracking time allows more flexibility to implement other commands
such as find free time or add recurring task.

Let's,
* add time support for deadline tasks

tyanhan committed:
Add recurring tag functionality and time labels on task

There are no recurring task tags.

Having these tags will allow users to differentiate their recurring
tasks and non recurring tasks.

Let's add the time labels and recurring tag functionality.

tyanhan committed:
Fix tests

tyanhan committed:
Add task recurring functionality

The task recurring labels have been added but the task isDone status
does not reset after the given timeframe.

Let's add the task reset functionality after a given timeframe.

tyanhan committed:
Merge pull request #4 from tyanhan/branch-B-RecurringTasks

Add recurring task functionality

tyanhan committed:
Add JavaDocs

tyanhan committed:
Change class names and create tag interface

tyanhan committed:
Improve code consistency

tyanhan committed:
Modify GUI, add support for resetting deadlines

tyanhan committed:
Add Ui screenshot

tyanhan committed:
Update README.md

tyanhan committed:
Update README.md

tyanhan committed:
Add logo and fix new Deadline exceeding date bug

tyanhan committed:
Merge branch 'master' of https://github.com/tyanhan/ip

tyanhan committed:
Change version number

tyanhan committed:
Add JavaDoc and reduced magic literals

tyanhan committed:
Fix User Guide

tyanhan committed:
Update User Guide

tyanhan committed:
Update User Guide

tyanhan committed:
Update User Guide

tyanhan committed:
Enable multispacing between command arguments

tyanhan committed:
Fix close app

tyanhan committed:
Fix error message on exit for MAC users

tyanhan committed:
Update README.md

tyanhan committed:
Update README.md
